tests/integration/test_shared_db.py

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/env python3
+# Copyright 2022 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import asyncio
+import logging
+
+import pytest
+from pytest_operator.plugin import OpsTest
+
+logger = logging.getLogger(__name__)
+
+MYSQL_APP_NAME = "mysql"
+KEYSTONE_APP_NAME = "keystone"
+MYSQLROUTER_APP_NAME = "mysqlrouter"
+
+
+@pytest.mark.order(1)
+@pytest.mark.abort_on_fail
+@pytest.mark.shared_db_tests
+async def test_shared_db(ops_test: OpsTest):
+    """Test the shared-db legacy relation."""
+    charm = await ops_test.build_charm(".")
+
+    mysql_app = await ops_test.model.deploy(
+        "mysql", channel="latest/edge", application_name=MYSQL_APP_NAME, num_units=3
+    )
+    keystone_app = await ops_test.model.deploy(
+        "keystone", application_name=KEYSTONE_APP_NAME, series="focal", num_units=2
+    )
+    mysqlrouter_app = await ops_test.model.deploy(
+        charm, application_name=MYSQLROUTER_APP_NAME, num_units=None
```

Comment:
As discussed on MM, this is slightly confusing - at first glance it looks like you're deploying a charm with no units. Perhaps add a comment describing that this is a subordinate charm, and therefore it'll scale automatically with keystone?